firebase-firestore/src/main/java/com/google/firebase/firestore/util/ProtoValueUtil.java

```diff
@@ -0,0 +1,266 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.util;
+
+import static com.google.firebase.firestore.util.Assert.fail;
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import androidx.annotation.Nullable;
+import com.google.common.base.Splitter;
+import com.google.firebase.firestore.model.value.FieldValue;
+import com.google.firestore.v1.ArrayValue;
+import com.google.firestore.v1.MapValue;
+import com.google.firestore.v1.Value;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class ProtoValueUtil {
+
+  public static int typeOrder(Value value) {
+
+    switch (value.getValueTypeCase()) {
+      case NULL_VALUE:
+        return FieldValue.TYPE_ORDER_NULL;
+      case BOOLEAN_VALUE:
+        return FieldValue.TYPE_ORDER_BOOLEAN;
+      case INTEGER_VALUE:
+        return FieldValue.TYPE_ORDER_NUMBER;
+      case DOUBLE_VALUE:
+        return FieldValue.TYPE_ORDER_NUMBER;
+      case TIMESTAMP_VALUE:
+        return FieldValue.TYPE_ORDER_TIMESTAMP;
+      case STRING_VALUE:
+        return FieldValue.TYPE_ORDER_STRING;
+      case BYTES_VALUE:
+        return FieldValue.TYPE_ORDER_BLOB;
+      case REFERENCE_VALUE:
+        return FieldValue.TYPE_ORDER_REFERENCE;
+      case GEO_POINT_VALUE:
+        return FieldValue.TYPE_ORDER_GEOPOINT;
+      case ARRAY_VALUE:
+        return FieldValue.TYPE_ORDER_ARRAY;
+      case MAP_VALUE:
+        return FieldValue.TYPE_ORDER_OBJECT;
+      default:
+        throw fail("Invalid value type: " + value.getValueTypeCase());
+    }
+  }
+
+  /** Returns whether `value` is non-null and corresponds to the given type order. */
+  public static boolean isType(@Nullable Value value, int typeOrder) {
+    return value != null && typeOrder(value) == typeOrder;
+  }
+
+  public static boolean equals(Value left, Value right) {
+    int leftType = typeOrder(left);
+    int rightType = typeOrder(right);
+    if (leftType != rightType) {
+      return false;
+    }
+
+    switch (leftType) {
+      case FieldValue.TYPE_ORDER_ARRAY:
+        return arrayEquals(left, right);
+      case FieldValue.TYPE_ORDER_NUMBER:
```

Comment:
Nit: consider ordering the switch cases consistently so that it's easier to audit/reason about differences (in this case, that would mean number, array, object).

Reply:
Done (here and everywhere). I am following the type order constants.